pkg/openstack/ironic.go

```diff
@@ -29,9 +34,96 @@ func ReconcileIronic(ctx context.Context, instance *corev1beta1.OpenStackControl
 			return res, err
 		}
 		instance.Status.Conditions.Remove(corev1beta1.OpenStackControlPlaneIronicReadyCondition)
+		instance.Status.Conditions.Remove(corev1beta1.OpenStackControlPlaneExposeIronicReadyCondition)
 		return ctrl.Result{}, nil
 	}
 
+	// add selector to service overrides
+	for _, endpointType := range []service.Endpoint{service.EndpointPublic, service.EndpointInternal} {
+		if instance.Spec.Ironic.Template.IronicAPI.Override.Service == nil {
+			instance.Spec.Ironic.Template.IronicAPI.Override.Service = map[service.Endpoint]service.RoutedOverrideSpec{}
+		}
+		instance.Spec.Ironic.Template.IronicAPI.Override.Service[endpointType] =
+			AddServiceComponentLabel(
+				instance.Spec.Ironic.Template.IronicAPI.Override.Service[endpointType],
+				ironic.Name+"-api")
+
+		if instance.Spec.Ironic.Template.IronicInspector.Override.Service == nil {
+			instance.Spec.Ironic.Template.IronicInspector.Override.Service = map[service.Endpoint]service.RoutedOverrideSpec{}
+		}
+		instance.Spec.Ironic.Template.IronicInspector.Override.Service[endpointType] =
+			AddServiceComponentLabel(
+				instance.Spec.Ironic.Template.IronicInspector.Override.Service[endpointType],
+				ironic.Name+"-inspector")
+	}
+
+	// When component services got created check if there is the need to create a route
+	if err := helper.GetClient().Get(ctx, types.NamespacedName{Name: "ironic", Namespace: instance.Namespace}, ironic); err != nil {
+		if !k8s_errors.IsNotFound(err) {
+			return ctrl.Result{}, err
+		}
+	}
+
+	// Ironic API
+	if ironic.Status.Conditions.IsTrue(ironicv1.IronicAPIReadyCondition) {
+		svcs, err := service.GetServicesListWithLabel(
+			ctx,
+			helper,
+			instance.Namespace,
+			map[string]string{common.AppSelector: ironic.Name + "-api"},
+		)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+
+		var ctrlResult reconcile.Result
+		instance.Spec.Ironic.Template.IronicAPI.Override.Service, ctrlResult, err = EnsureRoute(
+			ctx,
+			instance,
+			helper,
+			ironic,
+			svcs,
+			instance.Spec.Ironic.Template.IronicAPI.Override.Service,
+			instance.Spec.Ironic.APIOverride.Route,
+			corev1beta1.OpenStackControlPlaneExposeIronicReadyCondition,
+		)
+		if err != nil {
+			return ctrlResult, err
+		} else if (ctrlResult != ctrl.Result{}) {
+			return ctrlResult, nil
+		}
+	}
+
+	// Ironic Inspector
+	if ironic.Status.Conditions.IsTrue(ironicv1.IronicInspectorReadyCondition) {
```

Comment:
This condition is currently set to True by ironic-operator also when Inspector service is not requested (by setting replicas = 0). Would the remaining code here fail in that case? Or will `svcs` simply be empty an `EnsureRoute` would simply not create any resources?

Setting the condition to `True` when service is not requested in ironic-operator might not be the right thing to do ...

Reply:
when there is no service created by the ironic operator which has the IngressCreate annotation [1] (done by the service operator), there won't be a route created.

[1] https://github.com/openstack-k8s-operators/lib-common/blob/main/modules/common/service/types.go#L44